### PR TITLE
ARROW-7281: [C++] Make Adaptive builders' length match expectations

### PR DESCRIPTION
**Pull request by pitrou**
The length() value should account for internally buffered data, even if not committed to the final buffer.

### CODE REVIEW

**Comment by pitrou** [comment]
I've checked this doesn't affect the builder benchmarks.

**Review by bkietz — APPROVED** [review]
LGTM, nice catch

File: cpp/src/arrow/array/builder_adaptive.cc [hunk 1]
```diff
@@ -273,12 +283,16 @@ Status AdaptiveUIntBuilder::FinishInternal(std::shared_ptr<ArrayData>* out) {
...
   data_ = nullptr;
-  capacity_ = length_ = null_count_ = 0;
+  capacity_ = length_ = committed_length_ = null_count_ = 0;
   return Status::OK();
 }
 
 Status AdaptiveUIntBuilder::AppendValuesInternal(const uint64_t* values, int64_t length,
                                                  const uint8_t* valid_bytes) {
+  // UnsafeAppendToBitmap expects length_ to be the pre-update value, satisfy it
+  DCHECK_EQ(length_, committed_length_ + pending_pos_);
```
Comment (bkietz) [inline cpp/src/arrow/array/builder_adaptive.cc]: Could you delete committed_length_ by using `length_ -= pending_pos_` here?

Comment (pitrou) [reply cpp/src/arrow/array/builder_adaptive.cc]: You're right, that works. Thank you!

**Review by bkietz — CHANGES_REQUESTED** [review]
@pitrou [MSVC is barfing](https://ci.appveyor.com/project/pitrou/arrow/builds/30096622/job/0v80gfapipw070bn#L1168) on this implicit conversion (`u32 <- u64`) https://github.com/apache/arrow/pull/6174/files?file-filters%5B%5D=.cc#diff-346c62c75046920db82dd1a11c7893b3R1899

**Comment by pitrou** [comment]
Ok, hopefully it's fixed.

**Comment by pitrou** [comment]
AppVeyor: https://ci.appveyor.com/project/pitrou/arrow/builds/30108341

**Pull request closed without merging**